### DISCUSSION TITLE
How to integrate Firebase to iOS app

### DISCUSSION THREAD

**Discussion by crazyhappygame** [axmolengine/axmol#1817] [Q&A]
How to integrate Firebase to iOS?
I see multiple options here: https://firebase.google.com/docs/ios/installation-methods
- Standard pod install
- Swift Package Manager
- Installing from the GitHub repo

But that will not work nice with cmake. What is the recommended way for axmol?

**Comment by rh101** [comment]
While this isn't an Axmol issue, I've personally tried several methods of integrating Firebase, with the easiest being the Cocoa Pods method.  An example Podfile from one of my projects, this one for iOS/arm64, with project name changed:

```
project './build_ios_arm64/PROJECT_NAME.xcodeproj'

# Uncomment the next line to define a global platform for your project
platform :ios, '15.0'

target 'PROJECT_NAME' do
    inherit! :search_paths
    
    # Comment the next line if you don't want to use dynamic frameworks
    #use_frameworks! :linkage => :static
    use_frameworks!
    
    # Add the Firebase pod for Google Analytics
    pod 'FirebaseAnalytics'

    # Add the pods for any other Firebase products you want to use in your app
    pod 'FirebaseCrashlytics'
end
```
Run `pod init` once.

Then, after any update to your Xcode project files, such as via cmake or `axmol build` commands, re-run `pod install --repo-update`, and open the `PROJECT_NAME.xcworkspace` file in Xcode (not the `PROJECT_NAME.xcodeproj`).  That's pretty much it, and as cmake friendly as can be.

**Comment by smilediver** [comment]
We're using Pods so for now we have this setup for CMake:

```
        # Setup CocoaPods
        add_custom_target(LnPodfile
            COMMAND ${CMAKE_COMMAND} -E create_symlink
                ${CMAKE_CURRENT_SOURCE_DIR}/proj.ios_mac/Podfile
                ${CMAKE_CURRENT_BINARY_DIR}/Podfile
            DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/proj.ios_mac/Podfile
        )
        add_custom_target(LnPodfileLock
            COMMAND ${CMAKE_COMMAND} -E create_symlink
                ${CMAKE_CURRENT_SOURCE_DIR}/proj.ios_mac/Podfile.lock
                ${CMAKE_CURRENT_BINARY_DIR}/Podfile.lock
            DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/proj.ios_mac/Podfile.lock
        )
        add_custom_target(PodInstall
            COMMAND env PATH=\${PATH}:/opt/homebrew/bin LANG=en_US.UTF-8 bash -c \"pod install\"
            WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
            DEPENDS LnPodfile LnPodfileLock
        )
```

This symlinks `proj.ios_mac/Podfile` that is in the repo to the build directory, and adds `PodInstall` target, that you can run from Xcode to run this and instal pods.

The `proj.ios_mac/Podfile` has this to link to the project:

```
workspace '../../proj.ios_mac/GmsPlayground' # Adjust this depending to your project name and build path
project 'GmsPlayground'
```

**Comment by paulocoutinhox** [comment]
Hi,

1 - Create the file `Podfile.ios` in root directory with your content, example:

```ruby
# Settings
platform :ios, '12.0'
inhibit_all_warnings!
use_frameworks!

# Dependencies
target 'biblemania' do
    # Firebase
    pod 'FirebaseAnalytics'
    pod 'FirebaseCrashlytics'
end
```

2 - Add at the end of your CMake a command to copy podfile to build directory, example:

```
# podfile
if(APPLE)
    if(TVOS)
        set_target_properties(${APP_NAME} PROPERTIES XCODE_ATTRIBUTE_CONFIGURATION_BUILD_DIR "\${BUILD_DIR}/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)")
        configure_file("${CMAKE_CURRENT_SOURCE_DIR}/Podfile.tvos" "${CMAKE_CURRENT_BINARY_DIR}/Podfile" COPYONLY)
    elseif(IOS)
        set_target_properties(${APP_NAME} PROPERTIES XCODE_ATTRIBUTE_CONFIGURATION_BUILD_DIR "\${BUILD_DIR}/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)")
        configure_file("${CMAKE_CURRENT_SOURCE_DIR}/Podfile.ios" "${CMAKE_CURRENT_BINARY_DIR}/Podfile" COPYONLY)
    endif()
endif()
```

3 - Execute axmol build and pod install, example:

```
rm -rf build_ios_arm64/
axmol build -p ios -a arm64 -c
cd build_ios_arm64 && pod install
```

4 - Open your `xcworkspace` file always, example:

```
open ./build_ios_arm64/biblemania.xcworkspace/
```

Error in my case when i compile using Xcode:

```
ld: warning: ignoring duplicate libraries: '-lc++', '/Users/paulo/Developer/workspaces/cpp/biblemania/build_ios_arm64/engine/axmol/core/lib/libaxmol.a'
ld: framework 'Pods_biblemania' not found
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Can someone help me with this error?

```
framework 'Pods_biblemania' not found
```

But it is there:

<img width="1512" alt="image" src="https://github.com/axmolengine/axmol/assets/395096/29807603-9ab9-4242-ab63-0a8a1016dfa4">


Thanks.

**Reply by paulocoutinhox** [reply]
The solution is:

CMAKE:

```
if(APPLE)
    if(IOS)
        set_target_properties(${APP_NAME} PROPERTIES XCODE_ATTRIBUTE_CONFIGURATION_BUILD_DIR "\${BUILD_DIR}/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)")
        configure_file("${CMAKE_CURRENT_SOURCE_DIR}/Podfile.ios" "${CMAKE_CURRENT_BINARY_DIR}/Podfile" COPYONLY)
    endif()
endif()
```

PODFILE:

```
# Settings
platform :ios, '12.0'

# Dependencies
target 'biblemania' do
    inherit! :search_paths
    #use_frameworks! :linkage => :static # static link
    use_frameworks! # dynamic link

    # Firebase
    pod 'FirebaseAnalytics'
    pod 'FirebaseCrashlytics'
end
```

**Reply by meehjeevan** [reply]
FirebaseAnalytics is not working in mine.. 
Others are working.. 

I have pod like this:

```
target 'LudoX' do
  inherit! :search_paths
  use_frameworks!
  # Pods for LudoX
    pod 'FirebaseAnalytics'    
    pod 'FirebasePerformance'
    pod 'FirebaseCrashlytics'
    pod 'FirebaseRemoteConfig'
    pod 'GoogleUserMessagingPlatform'
    pod 'Google-Mobile-Ads-SDK'
end
```
FIRAnalytics is not working.. I'm getting this error:

`Use of undeclared identifier 'FIRAnalytics'`

But FirebaseAnalytics folder is there in pods. I tried cleaning/rebuilding but it didn't work.

**Reply by blairwang** [reply]
> framework 'Pods_biblemania' not found



> FirebaseAnalytics is not working in mine.. Others are working..
> 
> I have pod like this:
> 
> ```
> target 'LudoX' do
>   inherit! :search_paths
>   use_frameworks!
>   # Pods for LudoX
>     pod 'FirebaseAnalytics'    
>     pod 'FirebasePerformance'
>     pod 'FirebaseCrashlytics'
>     pod 'FirebaseRemoteConfig'
>     pod 'GoogleUserMessagingPlatform'
>     pod 'Google-Mobile-Ads-SDK'
> end
> ```
> 
> FIRAnalytics is not working.. I'm getting this error:
> 
> `Use of undeclared identifier 'FIRAnalytics'`
> 
> But FirebaseAnalytics folder is there in pods. I tried cleaning/rebuilding but it didn't work.

These two errors are same, I got this too.

fix using this may not good:
set_target_properties(${APP_NAME} PROPERTIES XCODE_ATTRIBUTE_CONFIGURATION_BUILD_DIR "\${BUILD_DIR}/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)")
        
I plan to try this:
        https://github.com/axmolengine/axmol/discussions/1162#discussioncomment-6173954
